OrbitLinuxTracing/ThreadStateVisitor.cpp

```diff
@@ -0,0 +1,272 @@
+  if (timestamp_ns < open_state.begin_timestamp_ns) {
+    // As noted above, overwrite the thread state retrieved at the beginning.
+    tid_open_states_.insert_or_assign(tid, OpenState{kNewState, timestamp_ns});
+    return std::nullopt;
+  }
+
+  if (open_state.state == kNewState || open_state.state == ThreadStateSlice::kRunning) {
+    // It seems to be somewhat common for a thread to receive a wakeup
+    // while already in runnable or running state: disregard the state change.
+    return std::nullopt;
+  }
+
+  if (open_state.state == ThreadStateSlice::kZombie ||
+      open_state.state == ThreadStateSlice::kDead) {
+    ERROR("Processed sched:sched_wakeup for thread %d but unexpected previous state %s", tid,
+          ThreadStateSlice::ThreadState_Name(open_state.state));
+  }
+
+  ThreadStateSlice slice;
+  slice.set_tid(tid);
+  slice.set_thread_state(open_state.state);
+  slice.set_begin_timestamp_ns(open_state.begin_timestamp_ns);
+  slice.set_end_timestamp_ns(timestamp_ns);
+  tid_open_states_.insert_or_assign(tid, OpenState{kNewState, timestamp_ns});
+  return slice;
+}
+
+std::optional<ThreadStateSlice> ThreadStateManager::OnContextSwitchIn(uint64_t timestamp_ns,
+                                                                      pid_t tid) {
+  constexpr ThreadStateSlice::ThreadState kNewState = ThreadStateSlice::kRunning;
+
+  auto open_state_it = tid_open_states_.find(tid);
+  if (open_state_it == tid_open_states_.end()) {
+    ERROR("Processed sched:sched_switch(in) but previous state of thread %d is unknown", tid);
+    tid_open_states_.insert_or_assign(tid, OpenState{kNewState, timestamp_ns});
+    return std::nullopt;
+  }
+
+  const OpenState& open_state = open_state_it->second;
+  if (timestamp_ns < open_state.begin_timestamp_ns) {
+    tid_open_states_.insert_or_assign(tid, OpenState{kNewState, timestamp_ns});
+    return std::nullopt;
+  }
+
+  if (open_state.state == kNewState) {
+    // No state change: do nothing and don't overwrite the previous begin timestamp.
+    return std::nullopt;
+  }
+
+  // Don't print an error even if open_state.state != ThreadStateSlice::kRunnable: it seems to be
+  // sometimes possible for a thread to go from a non-runnable state directly to running, skipping
+  // the sched:sched_wakeup event.
+
+  ThreadStateSlice slice;
+  slice.set_tid(tid);
+  slice.set_thread_state(open_state.state);
+  slice.set_begin_timestamp_ns(open_state.begin_timestamp_ns);
+  slice.set_end_timestamp_ns(timestamp_ns);
+  tid_open_states_.insert_or_assign(tid, OpenState{kNewState, timestamp_ns});
+  return slice;
+}
+
+std::optional<ThreadStateSlice> ThreadStateManager::OnContextSwitchOut(
+    uint64_t timestamp_ns, pid_t tid, ThreadStateSlice::ThreadState new_state) {
+  auto open_state_it = tid_open_states_.find(tid);
+  if (open_state_it == tid_open_states_.end()) {
+    ERROR("Processed sched:sched_switch(out) but previous state of thread %d is unknown", tid);
```

Comment:
(Also for the above): I would assume that this happens always on the beginning of the capture and is something usual. So maybe omit the error?

Reply:
Okay, then see above.